torch/nn/parallel/data_parallel.py

```diff
@@ -148,6 +148,16 @@ def forward(self, *inputs, **kwargs):
         if not self.device_ids:
             return self.module(*inputs, **kwargs)
 
+        # DataParallel needs at least one input in forward function
+        # Note:
+        # The forward function's input can be either in 'inputs' list or in 'kwargs' dict.
+        # If 'inputs' list is empty and 'kwargs' dict has item, we have no way to identify
+        # whether 'kwargs' has an input for forward function, so we just check whether `kwargs`
+        # dict has item. Caller needs to make sure 'kwargs' dict has at least an input for
+        # forward function in this case.
+        if len(inputs) == 0 and len(kwargs) == 0:
+            raise RuntimeError("Forward function must have at least one input, bot got zero")
```

Comment:
This looks OK to me.

But the use case mentioned in that issue looks valid to me, so I wonder if we should fix that use case instead of error out? 

Let me know what do you think, and whether you would like to fix. I can stamp this one if you think fixing is out of the scope. 

To fix, we need to following:

1. The segment fault occurs on `self.replicate`, we need to add the following to `replicate.py`:
    ```python
    if not devices:
        return []
    ```
2. After fixing that, we also need to update `parallel_apply.py` and `scatter_gather.py` accordingly to handle empty modules and empty outputs respectively.

Reply:
per our discussion, I will fix this entirely.